File: backend/workers/miro/src/board/create.py
import json

import grpc
import requests
from google.protobuf.struct_pb2 import Struct

from area_back_pb2 import JobError
from src.utils.parsing import get_arguments

from area_back_pb2_grpc import AreaBackServiceStub
from area_types_pb2 import JobData

URL = 'https://api.miro.com/v1/boards'
TARGET = "localhost:50050"


def create_board():
    args = get_arguments({"auth", "name", "description", "workflowStepId", "identifier"})
    target = args["target"] if args.keys().__contains__("target") else TARGET
    credentials = json.loads(args["auth"])

    headers = {
        'Authorization': f'Bearer {credentials["access_token"]}',
        'Content-Type': 'application/json'
    }
    board_data = {
        "name": args["name"],
        "description": args["description"],
        "teamId": credentials["team_id"],
    }

    response = requests.post(URL, headers=headers, json=board_data)

    if response.status_code == 201:
        created_board = response.json()
        with grpc.insecure_channel(target) as channel:
            params = Struct()
            params.update({
                "workflowStepId": args["workflowStepId"],
                "url": created_board["viewLink"],
                "createdAt": created_board["createdAt"],
            })
            AreaBackServiceStub(channel).OnReaction(
                JobData(name="miro-create-board", identifier=args["identifier"], params=params))

    else:
        error = response.json()
        with grpc.insecure_channel(target) as channel:
            AreaBackServiceStub(channel).OnError(
                JobError(identifier=args["identifier"], error=error["message"], isAuthError=(response.status_code == 401 or response.status_code == 403)))

File: backend/workers/twitter/src/tweet/create.ts
import { z } from "zod";
import { Client } from "twitter-api-sdk";
import { credentials } from "@grpc/grpc-js";
import { AreaBackServiceClient } from "../proto/area_back";
import "../proto/google/protobuf/struct";
import { onError, onReaction } from "../util/grpc";
import parseArguments, { TwitterAuthSchema } from "../util/params";
import { TwitterResponseError } from "../util/types";

const CreateTweetSchema = z.object({
	auth: TwitterAuthSchema,
	target: z.string().optional(),
	identifier: z.string(),
	workflowStepId: z.string(),
	text: z.string(),
});
type CreateTweetType = z.infer<typeof CreateTweetSchema>;

export default async function createTweet() {
	const params = parseArguments<CreateTweetType>(CreateTweetSchema);
	const client = new AreaBackServiceClient(params.target ?? "localhost:50050", credentials.createInsecure());

	try {
		const twitter = new Client(params.auth.access_token);

		await twitter.tweets.createTweet({
			text: params.text,
		});
		await onReaction(client, {
			name: "twitter-create-tweet",
			identifier: params.identifier,
			params: {
				workflowStepId: params.workflowStepId,
			},
		});
	} catch (e) {
		const error = e as TwitterResponseError;
		await onError(client, {
			identifier: params.identifier,
			error: error.error.detail,
			isAuthError: error.status === 401 || error.status === 403,
		});
	}
}
